apis/bigqueryconnection/v1beta1/connection_reference.go

```diff
@@ -44,46 +43,33 @@ func NewBigQueryConnectionConnectionRef(ctx context.Context, reader client.Reade
 	// Get location
 	location := obj.Spec.Location
 
-	// Get desired service-generated ID from spec
-	desiredServiceID := direct.ValueOf(obj.Spec.ResourceID)
-	if desiredServiceID != "" {
-		if _, err := uuid.Parse(desiredServiceID); err != nil {
-			return nil, fmt.Errorf("spec.resourceID should be in a UUID format, got %s ", desiredServiceID)
-		}
-	}
+	// Get desired connection ID from spec
+	desiredID := direct.ValueOf(obj.Spec.ResourceID)
 
-	// Get externalReference
+	// Validate status.externalRef
 	externalRef := direct.ValueOf(obj.Status.ExternalRef)
 	if externalRef != "" {
-		tokens := strings.Split(externalRef, "/")
-
-		if len(tokens) != 6 || tokens[0] != "projects" || tokens[2] != "locations" || tokens[4] != "connections" {
-			return nil, fmt.Errorf("externalRef should be projects/<project>/locations/<location>/connections/<Connection>, got %s", externalRef)
+		actualProject, actualLocation, actualID, err := parseExternal(externalRef)
+		if err != nil {
+			return nil, err
 		}
-		id.parent = "projects/" + tokens[1] + "/locations/" + tokens[3]
-
 		// Validate spec parent and resourceID field if the resource is already reconcilied with a GCP Connection resource.
-		if tokens[1] != projectID {
-			return nil, fmt.Errorf("spec.projectRef changed, expect %s, got %s",
-				tokens[1], projectID)
+		if projectID != actualProject {
+			return nil, fmt.Errorf("spec.projectRef changed, expect %s, got %s", projectID, actualProject)
 		}
-		if tokens[3] != location {
-			return nil, fmt.Errorf("spec.location changed, expect %s, got %s",
-				tokens[3], location)
+		if location != actualLocation {
+			return nil, fmt.Errorf("spec.location changed, expect %s, got %s", location, actualLocation)
 		}
-		if desiredServiceID != "" && tokens[5] != desiredServiceID {
-			// Service generated ID shall not be reset in the same BigQueryConnectionConnection.
+		if desiredID != "" && desiredID != actualID {
+			// Connection ID shall not be reset in the same BigQueryConnectionConnection.
 			// TODO: what if multiple BigQueryConnectionConnection points to the same GCP Connection?
 			return nil, fmt.Errorf("cannot reset `spec.resourceID` to %s, since it has already acquired the Connection %s",
-				desiredServiceID, tokens[5])
+				desiredID, actualID)
 		}
 		id.External = externalRef
 		return id, nil
 	}
-	id.parent = "projects/" + projectID + "/locations/" + location
-	if desiredServiceID != "" {
-		id.External = id.parent + "/connections/" + desiredServiceID
-	}
+	id.External = "projects/" + projectID + "/locations/" + location + "/connections/" + desiredID
```

Comment:
I don't follow this. The `external` could be an empty desiredID, is this intended?

Reply:
I see what this is doing now. It may be not clear for code readability.